File: 백준/Platinum/19580. 마스크가 필요해/마스크가 필요해.cc
#include <iostream>
#include <vector>
#include <map>
#include <set>
#include <algorithm>

using namespace std;

struct Citizen {
    long long L, R;
    Citizen(long long L, long long R) : L(L), R(R) {}
};

struct Shop {
    long long P;
    int X;
    Shop(long long P, int X) : P(P), X(X) {}
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N, M, X, answer = 0;
    long long L, R, P;

    vector<Citizen> citizens;
    vector<Shop> shops;

    set<long long> availableShops;
    map<long long, int> shopCounts;

    cin >> N >> M;

    for (int i = 0; i < N; i++) {
        cin >> L >> R;
        citizens.push_back({L, R});
    }

    for (int i = 0; i < M; i++) {
        cin >> P >> X;
        shops.push_back({P, X});
    }

    // 시민을 R 기준으로 정렬
    sort(citizens.begin(), citizens.end(), [](const Citizen& a, const Citizen& b) {
        return a.R < b.R;
    });

    // 상점을 P 기준으로 정렬
    sort(shops.begin(), shops.end(), [](const Shop& a, const Shop& b) {
        return a.P < b.P;
    });

    for (const auto& shop : shops) {
        availableShops.insert(shop.P);
        shopCounts[shop.P] += shop.X;
    }

    for (const auto& citizen : citizens)
    {
        auto it = availableShops.lower_bound(citizen.L); // L 이상인 상점 찾기

        if (it != availableShops.end() && *it <= citizen.R) // R 이하인지 확인
        {
            long long shopPrice = *it;
            ++answer;
            if (--shopCounts[shopPrice] == 0)
                availableShops.erase(shopPrice); // 상점 제거
        }
    }

    cout << answer;

    return 0;
}

File: 백준/Platinum/19580. 마스크가 필요해/마스크가 필요해.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        List<Citizen> citizens = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            long L = Long.parseLong(st.nextToken());
            long R = Long.parseLong(st.nextToken());
            citizens.add(new Citizen(L, R));
        }

        List<Shop> shops = new ArrayList<>();
        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            long P = Long.parseLong(st.nextToken());
            int X = Integer.parseInt(st.nextToken());
            shops.add(new Shop(P, X));
        }

        citizens.sort(Comparator.comparingLong(c -> c.R));
        shops.sort(Comparator.comparingLong(s -> s.P));

        TreeSet<Long> availableShops = new TreeSet<>();
        Map<Long, Integer> shopCounts = new HashMap<>();

        for (Shop shop : shops) {
            availableShops.add(shop.P);
            shopCounts.put(shop.P, shopCounts.getOrDefault(shop.P, 0) + shop.X);
        }

        int matched = 0;
        for (Citizen citizen : citizens) {
            Long shopPrice = availableShops.ceiling(citizen.L);

            if (shopPrice != null && shopPrice <= citizen.R) {
                matched++;
                int count = shopCounts.get(shopPrice);
                if (count == 1) {
                    availableShops.remove(shopPrice);
                }
                shopCounts.put(shopPrice, count - 1);
            }
        }

        System.out.println(matched);
    }

    static class Citizen {
        long L, R;

        Citizen(long L, long R) {
            this.L = L;
            this.R = R;
        }
    }

    static class Shop {
        long P;
        int X;

        Shop(long P, int X) {
            this.P = P;
            this.X = X;
        }
    }
}
